src/annotator/guest.js

```diff
@@ -130,6 +130,12 @@ export default class Guest {
     this._highlightsVisible = false;
     this._isAdderVisible = false;
 
+    // Lookup of annotation's `$tag` and associated anchors.
+    this._annotations = /** @type {Map<string, Anchor[]>} */ (new Map());
```

Comment:
After thinking about this a little longer, I think this can be simplified to maintaining one additional set which tracks the annotations that should be in the page. This can be updated 1) when an annotation is received, just before it is anchored and 2) when an annotation is detached. After anchoring completes, the logic can check if the annotation should still be present in the page and skip saving the anchor if not.

Simplifying the data structures like this reduces the possibility of bugs due them getting out of sync, and reduces the number of code paths that need to be exercised in tests.

```diff
diff --git a/src/annotator/guest.js b/src/annotator/guest.js
index 9b3dcb504..59e174fff 100644
--- a/src/annotator/guest.js
+++ b/src/annotator/guest.js
@@ -130,12 +130,6 @@ export default class Guest {
     this._highlightsVisible = false;
     this._isAdderVisible = false;
 
-    // Lookup of annotation's `$tag` and associated anchors.
-    this._annotations = /** @type {Map<string, Anchor[]>} */ (new Map());
-
-    // Tags of annotations deleted while being anchored.
-    this._deferredDetach = /** @type {Set<string>} */ (new Set());
-
     this._adder = new Adder(this.element, {
       onAnnotate: async () => {
         await this.createAnnotation();
@@ -169,6 +163,12 @@ export default class Guest {
      */
     this.anchors = [];
 
+    /**
+     * Tags of annotations that are currently anchored or being anchored in
+     * the guest.
+     */
+    this._annotations = /** @type {Set<string>} */ (new Set());
+
     // Set the frame identifier if it's available.
     // The "top" guest instance will have this as null since it's in a top frame not a sub frame
     /** @type {string|null} */
@@ -369,19 +369,13 @@ export default class Guest {
     this._sidebarRPC.on(
       'deleteAnnotation',
       /** @param {string} tag */
-      tag => {
-        if (this._annotations.has(tag)) {
-          this.detach(tag);
-        } else {
-          this._deferredDetach.add(tag);
-        }
-      }
+      tag => this.detach(tag)
     );
 
     this._sidebarRPC.on(
       'loadAnnotations',
       /** @param {AnnotationData[]} annotations */
-      annotations => annotations.forEach(annotation => this.anchor(annotation))
+      annotations => annotations.forEach(ann => this.anchor(ann))
     );
 
     // Discover and connect to the sidebar frame. All RPC events must be
@@ -391,8 +385,6 @@ export default class Guest {
   }
 
   destroy() {
-    this._annotations.clear();
-    this._deferredDetach.clear();
     this._portFinder.destroy();
     this._notifyGuestUnload();
     this._hypothesisInjector.destroy();
@@ -498,11 +490,19 @@ export default class Guest {
     // Remove existing anchors for this annotation.
     this.detach(annotation.$tag, false /* notify */);
 
+    this._annotations.add(annotation.$tag);
+
     // Resolve selectors to ranges and insert highlights.
     if (!annotation.target) {
       annotation.target = [];
     }
     const anchors = await Promise.all(annotation.target.map(locate));
+
+    // If the annotation was removed while anchoring, don't save the anchors.
+    if (!this._annotations.has(annotation.$tag)) {
+      return [];
+    }
+
     for (let anchor of anchors) {
       highlight(anchor);
     }
@@ -515,15 +515,10 @@ export default class Guest {
       anchors.every(anchor => anchor.target.selector && !anchor.range);
 
     this._updateAnchors(this.anchors.concat(anchors), true /* notify */);
-    this._annotations.set(annotation.$tag, anchors);
 
     // Let other frames (eg. the sidebar) know about the new annotation.
     this._sidebarRPC.call('syncAnchoringStatus', annotation);
 
-    if (this._deferredDetach.has(annotation.$tag)) {
-      this.detach(annotation.$tag);
-    }
-
     return anchors;
   }
 
@@ -534,6 +529,8 @@ export default class Guest {
    * @param {boolean} [notify] - For internal use. Whether to emit an `anchorsChanged` notification
    */
   detach(tag, notify = true) {
+    this._annotations.delete(tag);
+
     /** @type {Anchor[]} */
     const anchors = [];
     for (let anchor of this.anchors) {
@@ -543,10 +540,6 @@ export default class Guest {
         removeHighlights(anchor.highlights);
       }
     }
-
-    this._deferredDetach.delete(tag);
-    this._annotations.delete(tag);
-
     this._updateAnchors(anchors, notify);
   }
 
diff --git a/src/annotator/test/guest-test.js b/src/annotator/test/guest-test.js
index 02737f38f..50bb31e89 100644
--- a/src/annotator/test/guest-test.js
+++ b/src/annotator/test/guest-test.js
@@ -420,21 +420,18 @@ describe('Guest', () => {
     });
 
     describe('on "deleteAnnotation" event', () => {
-      it('defers detach annotation until it is anchored', async () => {
+      it('prevents saving of annotations that are currently anchoring', async () => {
         const guest = createGuest();
-        const callback = sinon.stub();
         const ann = { target: [], uri: 'uri', $tag: 'tag1' };
-        guest._emitter.subscribe('anchorsChanged', callback);
 
+        emitSidebarEvent('loadAnnotations', [ann]);
         emitSidebarEvent('deleteAnnotation', ann.$tag);
 
-        assert.notCalled(callback);
-
-        emitSidebarEvent('loadAnnotations', [ann]);
         await delay(0);
 
-        assert.calledOnce(callback);
-        assert.calledWith(callback, []);
+        assert.isUndefined(
+          guest.anchors.find(anchor => anchor.annotation === ann)
+        );
       });
     });
   });
```

Reply:
Another general comment I have is that if you have a set of fields which do need to be kept in sync, it is helpful to locate their definitions near each other so that someone else reading the code becomes aware of them.

Reply:
You are introducing a new data structure that can equally become out of sync, and with less future benefits. 

If you want to solve this without creating a new data structure, you should rely on `Guest#anchors`, and loop through the annotations.

My intention was to replace `Guest#anchors` by `Guest#annotations` (we can give the same name.